octodns_bind/__init__.py

```diff
@@ -124,22 +177,124 @@ def _load_zone_file(self, zone_name):
 
         return z
 
-    def zone_records(self, zone):
+    def zone_records(self, zone, target):
         if zone.name not in self._zone_records:
-            z = self._load_zone_file(zone.name)
+            z = self._load_zone_file(zone.name, target)
 
             records = []
-            for name, ttl, rdata in z.iterate_rdatas():
-                rdtype = dns.rdatatype.to_text(rdata.rdtype)
-                if rdtype in self.SUPPORTS:
-                    records.append(
-                        Rr(name.to_text(), rdtype, ttl, rdata.to_text())
-                    )
+            if z:
+                for name, ttl, rdata in z.iterate_rdatas():
+                    rdtype = dns.rdatatype.to_text(rdata.rdtype)
+                    if rdtype in self.SUPPORTS:
+                        records.append(
+                            Rr(name.to_text(), rdtype, ttl, rdata.to_text())
+                        )
 
             self._zone_records[zone.name] = records
 
         return self._zone_records[zone.name]
 
+    def _primary_nameserver(self, decoded_name, records):
+        for record in records:
+            if record.name == '' and record._type == 'NS':
+                return record.values[0]
+        self.log.warning(
+            '_primary_nameserver: unable to find a primary_nameserver for %s, using placeholder',
+            decoded_name,
+        )
+        return f'ns.{decoded_name}'
+
+    def _hostmaster_email(self, decoded_name):
+        pieces = self.hostmaster_email.split('@')
+        # escape any .'s in the email username
+        pieces[0] = pieces[0].replace('.', '\\.')
+        if len(pieces) == 2:
+            return '.'.join(pieces)
+
+        return f'{pieces[0]}.{decoded_name}'
+
+    def _longest_name(self, records):
+        try:
+            return sorted(len(r.name) for r in records)[-1]
+        except IndexError:
+            return 0
+
+    def _now(self):
+        return datetime.utcnow()
+
+    def _serial(self):
+        # things wrap/reset at max int
+        return int(self._now().timestamp()) % 2147483647
+
+    def _apply(self, plan):
+        desired = plan.desired
+        name = desired.decoded_name
+
+        if not isdir(self.directory):
+            makedirs(self.directory)
+
+        records = sorted(c.record for c in plan.changes)
+        longest_name = self._longest_name(records)
+
+        filename = join(self.directory, f'{name[:-1]}{self.file_extension}')
+        with open(filename, 'w') as fh:
+            template = Template(
+                '''$$ORIGIN $zone_name
+
+@ $default_ttl IN SOA $primary_nameserver $hostmaster_email (
+    $serial ; Serial
+    3600       ; Refresh (1 hour)
```

Comment:
Should these (Refresh, Retry, Expire, NXDOMAIN) be user configurable? Mine are actually the same for personal stuff, and I can't really think of a time I really touched those, but figured I'd ask your thoughts.

If they aren't, I'm fine with it, they are complete sane ones.